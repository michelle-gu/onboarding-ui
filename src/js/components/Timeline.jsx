import React, { Component } from "react";
import ReactDOM from "react-dom";
import Post from "./Post.jsx";

class Timeline extends Component {

  	constructor() {
    	super();
  	}

  	render() {
    	return (
      		<Post></Post>
    	);
  	}

}

export default Timeline;

ReactDOM.render(<Timeline />, document.getElementById("timeline-div"));
