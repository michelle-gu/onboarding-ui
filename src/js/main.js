const http = require('http');
const url = require('url');
const fs = require('fs');

const hostname = '127.0.0.1';
const port = 9000;

const server = http.createServer((req, res) => {
    var path = url.parse(req.url).pathname;  
    if (path === '/') { // Home page
        fs.readFile(__dirname + '/../index.html', function(error, data) {  
            if (error) {  
                res.writeHead(404);  
                res.write('File not found');  
            } else {  
                res.writeHead(200, {'Content-Type': 'text/html'});  
                res.write(data);  
            }
            res.end();    
        });  
    } else if (path.split('.')[1] == 'js') { // Javascript files
        fs.readFile(__dirname + path, function(error, data) {  
            if (error) {  
                res.writeHead(404);  
                res.write('File not found');  
            } else {  
                res.writeHead(200, {'Content-Type': 'application/javascript'});
                res.write(data);  
            }
            res.end();    
        });
    } else if (path.split('.')[1] == 'css') { // Javascript files
        fs.readFile(__dirname + '/..' + path, function(error, data) {  
            if (error) {  
                res.writeHead(404);  
                res.write('File not found');  
            } else {  
                res.writeHead(200, {'Content-Type': 'text/css'});
                res.write(data);  
            }
            res.end();    
        });   
    } else { // Page doesn't exist
        res.writeHead(404);  
        res.write("This page doesn't exist.");  
        res.end();  
    }
});

server.listen(port, hostname, () => {
    console.log(`Server running at http://${hostname}:${port}/`);
});

