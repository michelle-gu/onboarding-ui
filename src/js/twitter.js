// AJAX request for timeline
function getTimeline() {
  	var xhr = new XMLHttpRequest();
    xhr.open('GET', 'http://127.0.0.1:8080/api/1.0/twitter/timeline', true);
	xhr.onreadystatechange = function() {
		if (this.readyState == XMLHttpRequest.DONE && this.status == 200) {
			buildTimeline(this.responseText);
		}
	};
	xhr.onerror = function() {
    	console.log('Error making request to get timeline.');
		document.getElementById("timeline").innerHTML = "Unable to retrieve timeline at this time. Check back later.";
  	};
	xhr.send();
}

// Use JS to dynamically build timeline
function buildTimeline(responseText) {
	var jsonTimeline = JSON.parse(responseText).timeline;
	var timelineDiv = document.getElementById("timeline");
	timelineDiv.innerHTML = "";

	for (var i = 0; i < jsonTimeline.length; i++) {
		var post = jsonTimeline[i];

		// Create user info part of post
		var userDiv = document.createElement("div");
		userDiv.className = "user"

		var image = document.createElement("img");
		image.id = "profile-img";
		image.src = post.user.profileImageUrl;

		var nameSpan = document.createElement("span");
		nameSpan.className = "name";
		var name = document.createTextNode(post.user.name);
		nameSpan.appendChild(name);

		var screenNameSpan = document.createElement("span");
		screenNameSpan.className = "name small-text";
		var screenName = document.createTextNode(post.user.twitterHandle);
		screenNameSpan.appendChild(screenName);

		userDiv.appendChild(image);
		userDiv.appendChild(nameSpan);
		userDiv.appendChild(screenNameSpan);

		// Create tweet info part of post
		var tweetDiv = document.createElement("div");
		tweetDiv.className = "tweet"

		var timeDiv = document.createElement("div");
		timeDiv.className = "date small-text";
		var options = { month: 'short', day: 'numeric' };
		var createdAt  = new Date(post.createdAt);
		var time = document.createTextNode(createdAt.toLocaleDateString("en-US", options));
		timeDiv.appendChild(time);

		var messageDiv = document.createElement("div");
		var message = document.createTextNode(post.message);
		messageDiv.appendChild(message);

		tweetDiv.appendChild(timeDiv);
		tweetDiv.appendChild(messageDiv);

		// Link tweet to twitter
		var tweetLink = document.createElement("a");
		tweetLink.href = post.url;
		tweetLink.target = "_blank"
		tweetLink.appendChild(tweetDiv);

		// Create post
		var postDiv = document.createElement("div");
		postDiv.className = "post"
		postDiv.appendChild(userDiv);
		postDiv.appendChild(tweetLink);

		// Add post to timeline
		timelineDiv.appendChild(postDiv);
	}
}

window.onload = function () {
	getTimeline();
	document.getElementById("timeline-button").onClick = getTimeline();
}

