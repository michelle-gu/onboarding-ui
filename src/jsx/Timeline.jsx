import React, { Component } from "react";
import Post from "./Post.jsx";

class Timeline extends Component {

    constructor() {
      super();
    }

    generatePosts() {
      return this.props.timeline.map((post, index) => {
        let options = { month: 'short', day: 'numeric' };
        let createdAt  = new Date(post.createdAt).toLocaleDateString("en-US", options);
        return <Post key={index}
                     profileImageUrl={post.user.profileImageUrl}
                     name={post.user.name}
                     twitterHandle={post.user.twitterHandle}
                     createdAt={createdAt}
                     message={post.message}
                     url={post.url} />;
      });
    }

    render() {
      return (
        <div id="timeline">
          {this.generatePosts()}
        </div>
      );
    }

}

export default Timeline;
