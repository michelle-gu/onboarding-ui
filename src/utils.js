
// AJAX call to retrieve timeline
export function fetchTimeline() {
    return new Promise((resolve, reject) => {
        let xhr = new XMLHttpRequest();
        xhr.open('GET', 'http://127.0.0.1:8080/api/1.0/twitter/timeline', true);
        xhr.onreadystatechange = () => {
            if (xhr.readyState == XMLHttpRequest.DONE && xhr.status == 200) {
                resolve(JSON.parse(xhr.responseText).timeline);
            }
        };
        xhr.onerror = () => reject(new Error('Error making request to get timeline.'));
        xhr.send();
    });
}

export function fetchUserTimeline() {
    return new Promise((resolve, reject) => {
        let xhr = new XMLHttpRequest();
        xhr.open('GET', 'http://127.0.0.1:8080/api/1.0/twitter/user-timeline', true);
        xhr.onreadystatechange = () => {
            if (xhr.readyState == XMLHttpRequest.DONE && xhr.status == 200) {
                resolve(JSON.parse(xhr.responseText).timeline);
            }
        };
        xhr.onerror = () => reject(new Error('Error making request to get timeline.'));
        xhr.send();
    });
}
