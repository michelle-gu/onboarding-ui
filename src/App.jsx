import React, { Component } from 'react';
import Timeline from "./jsx/Timeline.jsx";

class App extends Component {

	constructor() {
		super();
	    this.state = {
	      timeline: [],
	    };
	}

	getTimeline() {
	  	let xhr = new XMLHttpRequest();
	    xhr.open('GET', 'http://127.0.0.1:8080/api/1.0/twitter/timeline', true);
		xhr.onreadystatechange = () => {
			if (xhr.readyState == XMLHttpRequest.DONE && xhr.status == 200) {
				let jsonTimeline = JSON.parse(xhr.responseText).timeline;
			    this.setState(() => {
			      return { timeline: jsonTimeline};
			    });
			}
		};
		xhr.onerror = () => {
	    	console.log('Error making request to get timeline.');
			document.getElementById("timeline").innerHTML = "Unable to retrieve timeline at this time. Check back later.";
	  	};
		xhr.send();
	}

	componentDidMount() {
		this.getTimeline();
	}
  
	render() {
		return (
			<div className="App">
				<header className="App-header title">Lab for Mich</header>

		        <div id="timeline-button-div">
		            <button id="timeline-button" className="button" type="button" onClick={() => this.getTimeline()}>Get Timeline</button>
		        </div>

		        <div id="timeline-div">
		        	<Timeline timeline={this.state.timeline} />
		        </div>
			</div>
		);
	}

}

export default App;
