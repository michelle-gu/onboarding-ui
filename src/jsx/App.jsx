import React, { Component } from 'react';
import Timeline from "./Timeline.jsx";
import { fetchTimeline } from '../utils.js';

class App extends Component {

    constructor(props) {
        super(props);
        this.state = {
            timeline: [],
        };

        this.updateTimeline = this.updateTimeline.bind(this);
    }

    updateTimeline() {
        fetchTimeline().then((jsonTimeline) => {
            this.setState({ timeline: jsonTimeline });
        }, (error) => {
            console.log(error);
            document.getElementById("timeline").innerHTML = "Unable to retrieve timeline at this time. Check back later.";
        });
    }

    componentDidMount() {
        this.updateTimeline();
    }

    render() {
        return (
            <div className="App">
                <header className="App-header title">Lab for Mich</header>

                <div id="timeline-button-div">
                    <button id="timeline-button" className="button" type="button" onClick={() => this.updateTimeline()} >Get Timeline</button>
                </div>

                <div id="timeline-div">
                    <Timeline timeline={this.state.timeline} />
                </div>
            </div>
        );
    }

}

export default App;
