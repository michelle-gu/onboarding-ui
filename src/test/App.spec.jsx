import React from 'react';
import { mount, shallow } from 'enzyme';
import App from '../jsx/App';
import Timeline from '../jsx/Timeline.jsx';
import { fetchTimeline } from '../utils.js'

const utils = require('../utils.js');

describe('App', () => {
    let wrapper;
    beforeEach(() => wrapper = shallow(<App />));

    it('should render correctly', () => expect(wrapper).toMatchSnapshot());

    it('should render 3 <div />', () => {
        expect(wrapper.find('div').length).toEqual(3);
    });

    it('should render 1 <header />', () => {
        expect(wrapper.find('header').length).toEqual(1);
    });

    it('should render 1 <button />', () => {
        expect(wrapper.find('button').length).toEqual(1);
    });

    it('should render the Timeline Component', () => {
        expect(wrapper.containsMatchingElement(<Timeline timeline={wrapper.instance().state.timeline} />)).toEqual(true);
    });

    it('test button click event', () => {
        utils.fetchTimeline = jest.fn();
        wrapper.find('button').simulate('click');
        expect(utils.fetchTimeline).toHaveBeenCalledTimes(1);
    });

    it('test `componentDidMount()`', () => {
        utils.fetchTimeline = jest.fn();
        wrapper.instance().componentDidMount();
        expect(utils.fetchTimeline).toHaveBeenCalledTimes(1);
    });

    it('test getTimeline function', () => {
        wrapper.instance().getTimeline([{"message": "test"}]);
        expect(wrapper.state('timeline')).toEqual([{"message": "test"}]);
    });

});

