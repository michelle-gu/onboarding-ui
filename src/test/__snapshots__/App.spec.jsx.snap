// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<div
  className="App"
>
  <header
    className="App-header title"
  >
    Lab for Mich
  </header>
  <div
    id="timeline-button-div"
  >
    <button
      className="button"
      id="timeline-button"
      onClick={[Function]}
      type="button"
    >
      Get Timeline
    </button>
  </div>
  <div
    id="timeline-div"
  >
    <Timeline
      timeline={Array []}
    />
  </div>
</div>
`;
